src/tools/miri/src/shims/x86/mod.rs

```diff
@@ -1000,23 +1000,26 @@ fn mask_store<'tcx>(
     mask: &OpTy<'tcx>,
     value: &OpTy<'tcx>,
 ) -> InterpResult<'tcx, ()> {
-    let (mask, mask_len) = this.operand_to_simd(mask)?;
-    let (value, value_len) = this.operand_to_simd(value)?;
+    let (mask, mask_len) = this.project_to_simd(mask)?;
+    let (value, value_len) = this.project_to_simd(value)?;
 
     assert_eq!(value_len, mask_len);
 
     let mask_item_size = mask.layout.field(this, 0).size;
     let high_bit_offset = mask_item_size.bits().strict_sub(1);
 
+    // Deref `ptr` at the `value` type (which is an array).
     let ptr = this.read_pointer(ptr)?;
     for i in 0..value_len {
         let mask = this.project_index(&mask, i)?;
         let value = this.project_index(&value, i)?;
 
         if this.read_scalar(&mask)?.to_uint(mask_item_size)? >> high_bit_offset != 0 {
+            // *Bon-inbounds* pointer arithmetic to compute the destination.
             let ptr = ptr.wrapping_offset(value.layout.size * i, &this.tcx);
-            // Unaligned copy, which is what we want.
-            this.mem_copy(value.ptr(), ptr, value.layout.size, /*nonoverlapping*/ true)?;
+            // Deref the pointer *unaligned*, and do the copy.
+            let dest = this.ptr_to_mplace_unaligned(ptr, value.layout);
+            this.copy_op(&value, &dest)?;
```

Comment:
I had to adjust the logic a bit here because `operand_to_simd` used to return an `MPlace` but `project_to_simd` preserves the original type.